BUG, TST: random/test/test_extending:test_cython tries to write in wrong directory

<!-- Please describe the issue in detail here, and fill in the fields below -->
One of the tests fails, the previous version of numpy + earlier release of icc worked fine.
### Reproducing code example:

<!-- A short code example that reproduces the problem/missing feature. It should be
self-contained, i.e., possible to run as-is via 'python myproblem.py' -->

```python
python -c 'import numpy; numpy.test("full");'
```


```
_________________________________________________________________________________________ TestKind.test_all __________________________________________________________________________________________

self = <numpy.f2py.tests.test_kind.TestKind object at 0x7f8856d5c9d0>

    @pytest.mark.slow
    def test_all(self):
        selectedrealkind = self.module.selectedrealkind
        selectedintkind = self.module.selectedintkind
    
        for i in range(40):
            assert_(selectedintkind(i) in [selected_int_kind(i), -1],
                    'selectedintkind(%s): expected %r but got %r' %
                    (i, selected_int_kind(i), selectedintkind(i)))
    
        for i in range(20):
>           assert_(selectedrealkind(i) in [selected_real_kind(i), -1],
                    'selectedrealkind(%s): expected %r but got %r' %
                    (i, selected_real_kind(i), selectedrealkind(i)))
E           AssertionError: selectedrealkind(16): expected 10 but got 16

i          = 16
selectedintkind = <fortran object>
selectedrealkind = <fortran object>
self       = <numpy.f2py.tests.test_kind.TestKind object at 0x7f8856d5c9d0>

/usr/lib/python3.8/site-packages/numpy/f2py/tests/test_kind.py:32: AssertionError
____________________________________________________________________________________________ test_cython _____________________________________________________________________________________________

    @pytest.mark.skipif(cython is None, reason="requires cython")
    def test_cython():
        curdir = os.getcwd()
        argv = sys.argv
        examples = (os.path.dirname(__file__), '..', '_examples')
        try:
            os.chdir(os.path.join(*examples))
            sys.argv = argv[:1] + ['build']
            with warnings.catch_warnings(record=True) as w:
                # setuptools issue gh-1885
                warnings.filterwarnings('always', '', DeprecationWarning)
>               from numpy.random._examples.cython import setup

argv       = ['-c']
curdir     = '/home/zerg/Work/python-numpy-mkl'
examples   = ('/usr/lib/python3.8/site-packages/numpy/random/tests', '..', '_examples')
w          = [<warnings.WarningMessage object at 0x7f884e7361f0>]

/usr/lib/python3.8/site-packages/numpy/random/tests/test_extending.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/setup.py:30: in <module>
    ext_modules=cythonize(extensions)
/usr/lib/python3.8/site-packages/Cython/Build/Dependencies.py:1096: in cythonize
    cythonize_one(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pyx_file = '/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx', c_file = '/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.c', fingerprint = None
quiet = False, options = <Cython.Compiler.Main.CompilationOptions object at 0x7f88540d5fa0>, raise_on_failure = True
embedded_metadata = {'distutils': {'depends': ['/usr/lib/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h', '/usr/lib/python...ources': ['/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx']}, 'module_name': 'extending'}
full_module_name = 'extending', progress = '[1/2] '

    @record_results
    def cythonize_one(pyx_file, c_file, fingerprint, quiet, options=None,
                      raise_on_failure=True, embedded_metadata=None, full_module_name=None,
                      progress=""):
        from ..Compiler.Main import compile_single, default_options
        from ..Compiler.Errors import CompileError, PyrexError
    
        if fingerprint:
            if not os.path.exists(options.cache):
                safe_makedirs(options.cache)
            # Cython-generated c files are highly compressible.
            # (E.g. a compression ratio of about 10 for Sage).
            fingerprint_file_base = join_path(
                options.cache, "%s-%s" % (os.path.basename(c_file), fingerprint))
            gz_fingerprint_file = fingerprint_file_base + gzip_ext
            zip_fingerprint_file = fingerprint_file_base + '.zip'
            if os.path.exists(gz_fingerprint_file) or os.path.exists(zip_fingerprint_file):
                if not quiet:
                    print("%sFound compiled %s in cache" % (progress, pyx_file))
                if os.path.exists(gz_fingerprint_file):
                    os.utime(gz_fingerprint_file, None)
                    with contextlib.closing(gzip_open(gz_fingerprint_file, 'rb')) as g:
                        with contextlib.closing(open(c_file, 'wb')) as f:
                            shutil.copyfileobj(g, f)
                else:
                    os.utime(zip_fingerprint_file, None)
                    dirname = os.path.dirname(c_file)
                    with contextlib.closing(zipfile.ZipFile(zip_fingerprint_file)) as z:
                        for artifact in z.namelist():
                            z.extract(artifact, os.path.join(dirname, artifact))
                return
        if not quiet:
            print("%sCythonizing %s" % (progress, pyx_file))
        if options is None:
            options = CompilationOptions(default_options)
        options.output_file = c_file
        options.embedded_metadata = embedded_metadata
    
        any_failures = 0
        try:
            result = compile_single(pyx_file, options, full_module_name=full_module_name)
            if result.num_errors > 0:
                any_failures = 1
        except (EnvironmentError, PyrexError) as e:
            sys.stderr.write('%s\n' % e)
            any_failures = 1
            # XXX
            import traceback
            traceback.print_exc()
        except Exception:
            if raise_on_failure:
                raise
            import traceback
            traceback.print_exc()
            any_failures = 1
        if any_failures:
            if raise_on_failure:
>               raise CompileError(None, pyx_file)
E               Cython.Compiler.Errors.CompileError: /usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx

CompileError = <class 'Cython.Compiler.Errors.CompileError'>
PyrexError = <class 'Cython.Compiler.Errors.PyrexError'>
any_failures = 1
c_file     = '/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.c'
compile_single = <function compile_single at 0x7f884fef6310>
default_options = {'annotate': None, 'annotate_coverage_xml': None, 'build_dir': None, 'c_line_in_traceback': True, ...}
embedded_metadata = {'distutils': {'depends': ['/usr/lib/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h', '/usr/lib/python...ources': ['/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx']}, 'module_name': 'extending'}
fingerprint = None
full_module_name = 'extending'
options    = <Cython.Compiler.Main.CompilationOptions object at 0x7f88540d5fa0>
progress   = '[1/2] '
pyx_file   = '/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx'
quiet      = False
raise_on_failure = True
traceback  = <module 'traceback' from '/usr/lib/python3.8/traceback.py'>

/usr/lib/python3.8/site-packages/Cython/Build/Dependencies.py:1219: CompileError
---------------------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------------------
Compiling /usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx because it changed.
Compiling /usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending_distributions.pyx because it changed.
[1/2] Cythonizing /usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.pyx
---------------------------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------------------------
[Errno 13] Permission denied: '/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.c'
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/Cython/Build/Dependencies.py", line 1202, in cythonize_one
    result = compile_single(pyx_file, options, full_module_name=full_module_name)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/Main.py", line 727, in compile_single
    return run_pipeline(source, options, full_module_name)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/Main.py", line 515, in run_pipeline
    err, enddata = Pipeline.run_pipeline(pipeline, source)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/Pipeline.py", line 355, in run_pipeline
    data = run(phase, data)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/Pipeline.py", line 335, in run
    return phase(data)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/Pipeline.py", line 52, in generate_pyx_code_stage
    module_node.process_implementation(options, result)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/ModuleNode.py", line 143, in process_implementation
    self.generate_c_code(env, options, result)
  File "/usr/lib/python3.8/site-packages/Cython/Compiler/ModuleNode.py", line 411, in generate_c_code
    f = open_new_file(result.c_file)
  File "/usr/lib/python3.8/site-packages/Cython/Utils.py", line 75, in open_new_file
    return codecs.open(path, "w", encoding="ISO-8859-1")
  File "/usr/lib/python3.8/codecs.py", line 905, in open
    file = builtins.open(filename, mode, buffering)
PermissionError: [Errno 13] Permission denied: '/usr/lib/python3.8/site-packages/numpy/random/_examples/cython/extending.c'
2 failed, 10399 passed, 85 skipped, 17 xfailed in 340.77s (0:05:40)
```


### Numpy/Python version information:

<!-- Output from 'import sys, numpy; print(numpy.__version__, sys.version)' -->
1.18.0 3.8.0 (default, Oct 23 2019, 18:51:26) 
[GCC 9.2.0]

--- a/numpy/random/_examples/cython/setup.py
+++ b/numpy/random/_examples/cython/setup.py
@@ -12,6 +12,7 @@ from setuptools.extension import Extension
 from os.path import join, abspath, dirname
 
 path = abspath(dirname(__file__))
+defs = [('NPY_NO_DEPRECATED_API', 0)]
 
 extending = Extension("extending",
                       sources=[join(path, 'extending.pyx')],
@@ -19,10 +20,13 @@ extending = Extension("extending",
                             np.get_include(),
                             join(path, '..', '..')
                         ],
+                      define_macros=defs,
                       )
 distributions = Extension("extending_distributions",
                           sources=[join(path, 'extending_distributions.pyx')],
-                          include_dirs=[np.get_include()])
+                          include_dirs=[np.get_include()],
+                          define_macros=defs,
+                         )
 
 extensions = [extending, distributions]
 

--- a/numpy/random/tests/test_extending.py
+++ b/numpy/random/tests/test_extending.py
@@ -26,18 +26,16 @@ try:
     import cython
 except ImportError:
     cython = None
-
-cython_ver = cython.__version__.split('.')
-if len(cython_ver) < 3 or cython_ver < ['0', '29', '14']:
-    # too old or wrong cython, skip the test
-    cython = None
+else:
+    cython_ver = cython.__version__.split('.')
+    if len(cython_ver) < 3 or cython_ver < ['0', '29', '14']:
+        # too old or wrong cython, skip the test
+        cython = None
 
 @pytest.mark.skipif(cython is None, reason="requires cython")
 @pytest.mark.slow
 @pytest.mark.skipif(sys.platform == 'win32', reason="cmd too long on CI")
 def test_cython(tmp_path):
-    curdir = os.getcwd()
-    argv = sys.argv
     examples = os.path.join(os.path.dirname(__file__), '..', '_examples')
     base = os.path.dirname(examples)
     shutil.copytree(examples, tmp_path / '_examples')

--- a/numpy/random/tests/test_extending.py
+++ b/numpy/random/tests/test_extending.py
@@ -1,6 +1,8 @@
 import os, sys
 import pytest
 import warnings
+import shutil
+import subprocess
 
 try:
     import cffi
@@ -25,21 +27,22 @@ try:
 except ImportError:
     cython = None
 
+cython_ver = cython.__version__.split('.')
+if len(cython_ver) < 3 or cython_ver < ['0', '29', '14']:
+    # too old or wrong cython, skip the test
+    cython = None
+
 @pytest.mark.skipif(cython is None, reason="requires cython")
-def test_cython():
+@pytest.mark.slow
+def test_cython(tmp_path):
     curdir = os.getcwd()
     argv = sys.argv
-    examples = (os.path.dirname(__file__), '..', '_examples')
-    try:
-        os.chdir(os.path.join(*examples))
-        sys.argv = argv[:1] + ['build']
-        with warnings.catch_warnings(record=True) as w:
-            # setuptools issue gh-1885
-            warnings.filterwarnings('always', '', DeprecationWarning)
-            from numpy.random._examples.cython import setup
-    finally:
-        sys.argv = argv
-        os.chdir(curdir)
+    examples = os.path.join(os.path.dirname(__file__), '..', '_examples')
+    base = os.path.dirname(examples)
+    shutil.copytree(examples, tmp_path / '_examples')
+    env = os.environ.copy()
+    subprocess.check_call([sys.executable, 'setup.py', 'build'], env=env,
+                          cwd=str(tmp_path / '_examples' / 'cython'))
 
 @pytest.mark.skipif(numba is None or cffi is None,
                     reason="requires numba and cffi")

--- a/numpy/random/tests/test_extending.py
+++ b/numpy/random/tests/test_extending.py
@@ -28,6 +28,10 @@ except ImportError:
     cython = None
 else:
     cython_ver = cython.__version__.split('.')
+    # Cython 0.29.14 is required for Python 3.8 and there are
+    # other fixes in the 0.29 series that are needed even for earlier
+    # Python versions.
+    # Note: keep in sync with the one in pyproject.toml
     if len(cython_ver) < 3 or cython_ver < ['0', '29', '14']:
         # too old or wrong cython, skip the test
         cython = None
@@ -39,8 +43,8 @@ def test_cython(tmp_path):
     examples = os.path.join(os.path.dirname(__file__), '..', '_examples')
     base = os.path.dirname(examples)
     shutil.copytree(examples, tmp_path / '_examples')
-    env = os.environ.copy()
-    subprocess.check_call([sys.executable, 'setup.py', 'build'], env=env,
+    subprocess.check_call([sys.executable, 'setup.py', 'build'],
+                          env=os.environ,
                           cwd=str(tmp_path / '_examples' / 'cython'))
 
 @pytest.mark.skipif(numba is None or cffi is None,

--- a/numpy/random/tests/test_extending.py
+++ b/numpy/random/tests/test_extending.py
@@ -34,6 +34,7 @@ if len(cython_ver) < 3 or cython_ver < ['0', '29', '14']:
 
 @pytest.mark.skipif(cython is None, reason="requires cython")
 @pytest.mark.slow
+@pytest.mark.skipif(sys.platform == 'win32', reason="cmd too long on CI")
 def test_cython(tmp_path):
     curdir = os.getcwd()
     argv = sys.argv
@@ -47,8 +48,8 @@ def test_cython(tmp_path):
 @pytest.mark.skipif(numba is None or cffi is None,
                     reason="requires numba and cffi")
 def test_numba():
-        from numpy.random._examples.numba import extending
+    from numpy.random._examples.numba import extending
 
 @pytest.mark.skipif(cffi is None, reason="requires cffi")
 def test_cffi():
-        from numpy.random._examples.cffi import extending
+    from numpy.random._examples.cffi import extending

--- a/numpy/random/tests/test_extending.py
+++ b/numpy/random/tests/test_extending.py
@@ -24,15 +24,17 @@ except ImportError:
 
 try:
     import cython
+    from Cython.Compiler.Version import version as cython_version
 except ImportError:
     cython = None
 else:
-    cython_ver = cython.__version__.split('.')
+    from distutils.version import LooseVersion
     # Cython 0.29.14 is required for Python 3.8 and there are
     # other fixes in the 0.29 series that are needed even for earlier
     # Python versions.
     # Note: keep in sync with the one in pyproject.toml
-    if len(cython_ver) < 3 or cython_ver < ['0', '29', '14']:
+    required_version = LooseVersion('0.29.14')
+    if LooseVersion(cython_version) < required_version:
         # too old or wrong cython, skip the test
         cython = None
 
@@ -44,7 +46,6 @@ def test_cython(tmp_path):
     base = os.path.dirname(examples)
     shutil.copytree(examples, tmp_path / '_examples')
     subprocess.check_call([sys.executable, 'setup.py', 'build'],
-                          env=os.environ,
                           cwd=str(tmp_path / '_examples' / 'cython'))
 
 @pytest.mark.skipif(numba is None or cffi is None,

